### PR TITLE
database, storage_proxy: Reconcile pages with dead rows and partitions incrementally

### PR DESCRIPTION
**Pull request by denesb**
Currently, mutation query on replica side will not respond with a result which doesn't have at least one live row. This causes problems if there is a lot of dead rows or partitions before we reach a live row, which stem from the fact that resulting reconcilable_result will be large:

1. Large allocations.  Serialization of reconcilable_result causes large allocations for storing result rows in std::deque
2. Reactor stalls. Serialization of reconcilable_result on the replica side and on the coordinator side causes reactor stalls. This impacts not only the query at hand. For 1M dead rows, freezing takes 130ms, unfreezing takes 500ms. Coordinator  does multiple freezes and unfreezes. The reactor stall on the coordinator side is >5s
3. Too large repair mutations. If reconciliation works on large pages, repair may fail due to too large mutation size. 1M dead rows is already too much: Refs https://github.com/scylladb/scylladb/issues/9111.

This patch fixes all of the above by making mutation reads respect the memory accounter's limit for the page size, even for dead rows. 

This patch also addresses the problem of client-side timeouts during paging. Reconciling queries processing long strings of tombstones will now properly page tombstones,like regular queries do.

My testing shows that this solution even increases efficiency. I tested with a cluster of 2 nodes, and a table of RF=2. The data layout was as follows (1 partition):
* Node1: 1 live row, 1M dead rows
* Node2: 1M dead rows, 1 live row

This was designed to trigger reconciliation right from the very start of the query.

Before:
```
Running query (node2, CL=ONE, cold cache)
Query done, duration: 140.0633503ms, pages: 101, result: [Row(pk=0, ck=3000000, v=0)]
Running query (node2, CL=ONE, hot cache)
Query done, duration: 66.7195275ms, pages: 101, result: [Row(pk=0, ck=3000000, v=0)]
Running query (all-nodes, CL=ALL, reconcile, cold-cache)
Query done, duration: 873.5400742ms, pages: 2, result: [Row(pk=0, ck=0, v=0), Row(pk=0, ck=3000000, v=0)]
```

After:
```
Running query (node2, CL=ONE, cold cache)
Query done, duration: 136.9035122ms, pages: 101, result: [Row(pk=0, ck=3000000, v=0)]
Running query (node2, CL=ONE, hot cache)
Query done, duration: 69.5286021ms, pages: 101, result: [Row(pk=0, ck=3000000, v=0)]
Running query (all-nodes, CL=ALL, reconcile, cold-cache)
Query done, duration: 162.6239498ms, pages: 100, result: [Row(pk=0, ck=0, v=0), Row(pk=0, ck=3000000, v=0)]
```

Non-reconciling queries have almost identical duration (1 few ms changes can be observed between runs). Note how in the after case, the reconciling read also produces 100 pages, vs. just 2 pages in the before case, leading to a much lower duration (less than 1/4 of the before).

Refs https://github.com/scylladb/scylladb/issues/7929
Refs https://github.com/scylladb/scylladb/issues/3672
Refs https://github.com/scylladb/scylladb/issues/7933
Fixes https://github.com/scylladb/scylladb/issues/9111

### CODE REVIEW

**Comment by denesb** [comment]
This PR is a resurrection of https://github.com/scylladb/scylladb/pull/9223.
I have rebased it on master and dropped unnecessary code: empty page handling in CQL layer, this was already done by the empty replica pages effort  - https://github.com/scylladb/scylladb/pull/11053.
I have also renamed the slice option from `allow_empty_pages` to `allow_mutation_read_page_without_live_row` as `allow_empty_pages` is too ambiguous with the feature introduced by https://github.com/scylladb/scylladb/pull/11053 (it even introduced a gossip feature called `ALLOW_EMPTY_PAGES`).

I still have to redo the testing.

**Comment by denesb** [comment]
I started thinking on how to test this. @tgrabiec did the following:

> I tested with a partition which has in clustering order 1M row tombstones followed by live row on one replica, and on then a live row on another replica. So that old data query would enter reconciliation

This case would not trigger a huge reconciliation anymore (will confirm to be sure), because both replicas would serve empty pages after 10K dead rows. When they get to the live rows, these rows will be behind at most 10K-1 dead rows. So the worst case scenario for read-repair is 10K tombstone + some data. The effects of this PR will be much less dramatic in a case like that, to the point where I started to question if this PR is even needed anymore. I will redo the test-case to confirm.

**Comment by tgrabiec** [comment]
What if you have a digest mismatch in the first data query page, wouldn't it trigger large reconciliation?

**Comment by tgrabiec** [comment]
When only one replica has a live row.

**Comment by avikivity** [comment]
> 
> My testing shows that this solution significantly reduces efficiency. I tested 1M dead rows, which takes 80MB as a single frozen mutation.
> 
> cl=one, no reconciliation, hot cache: 460 ms cl=two, reconciling, no paging (before the patch), hot cache: 6.2 s cl=two, reconciling, coordinator-side paging (after the patch), hot cache: 24.3 s
> 
> The number of page requests in that test was about 80.


Did it reduce efficiency, or just latency/throughput?

Efficiency = rows per instruction
Latency = time for last row
Throughput = rows per unit time

I can understand it somewhat reducing efficiency and significantly impacting latency.

**Comment by avikivity** [comment]
> Currently, mutation query on replica side will not respond with a result which doesn't have at least one live row. This causes problems if there is a lot of dead rows or partitions before we reach a live row, which stem from the fact that resulting reconcilable_result will be large:


This "Currently" refers to the past, yes? Maybe incremental reconciliation is indeed not necessary.

**Comment by denesb** [comment]
> > My testing shows that this solution significantly reduces efficiency. I tested 1M dead rows, which takes 80MB as a single frozen mutation.
> > cl=one, no reconciliation, hot cache: 460 ms cl=two, reconciling, no paging (before the patch), hot cache: 6.2 s cl=two, reconciling, coordinator-side paging (after the patch), hot cache: 24.3 s
> > The number of page requests in that test was about 80.
> 
> Did it reduce efficiency, or just latency/throughput?
> 
> Efficiency = rows per instruction Latency = time for last row Throughput = rows per unit time
> 
> I can understand it somewhat reducing efficiency and significantly impacting latency.

For I just copy+pasted @tgrabiec's cover letter. I will update it once I did my own testing (in-progress).

**Comment by denesb** [comment]
> 



> > Currently, mutation query on replica side will not respond with a result which doesn't have at least one live row. This causes problems if there is a lot of dead rows or partitions before we reach a live row, which stem from the fact that resulting reconcilable_result will be large:
> 
> This "Currently" refers to the past, yes? Maybe incremental reconciliation is indeed not necessary.

This is still the case. We only fixed data-queries.

**Comment by avikivity** [comment]
Maybe we should make mutation queries return pages without live rows (they wouldn't be empty). Doesn't that fix the problem?

**Comment by avikivity** [comment]
Note some efficiency loss on reconciliation isn't too bad, since it's supposed to be a rare operation.

**Comment by denesb** [comment]
> Maybe we should make mutation queries return pages without live rows (they wouldn't be empty). Doesn't that fix the problem?

That is exactly what this PR does.

File: mutation/mutation_partition.cc [hunk 1]
```diff
@@ -2220,7 +2220,7 @@ stop_iteration reconcilable_result_builder::consume(clustering_row&& cr, row_tom
     }
     _live_rows += is_alive;
```
Comment (avikivity) [inline mutation/mutation_partition.cc:2222]: > This pach doesn't address the problem of client-side timeouts during
paging. It is an orthogonal problem. Query pager could periodically
send empty pages to the client, or use some other keep-alive
mechanism.

This implies the coordinator loops around the reconciled pages. But why? Just send an empty result to the caller.

As long as the continue-where cookie makes forward progress, the client will ask for another page, and we'll make forward progress.

File: service/storage_proxy.cc [hunk 2]
```diff
@@ -4984,9 +4986,11 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
 
                 // We generate a retry if at least one node reply with count live columns but after merge we have less
                 // than the total number of column we are interested in (which may be < count on a retry).
-                // So in particular, if no host returned count live columns, we know it's not a short read.
-                bool can_send_short_read = rr_opt && rr_opt->is_short_read() && rr_opt->row_count() > 0;
-                if (rr_opt && (can_send_short_read || data_resolver->all_reached_end() || rr_opt->row_count() >= original_row_limit()
+                // So in particular, if no host returned count live columns, we know it's not a short read due to
+                // row or partition limits being exhausted and retry is not needed.
```
Comment (avikivity) [inline service/storage_proxy.cc:4990]: I don't fully understand the comment, and if I did, I probably wouldn't like it. "short read due to no data" vs "end of data, don't read more pages" should be explicit, not inferred from a complicated set of conditions.

File: mutation/mutation_partition.cc [hunk 3]
```diff
@@ -2200,6 +2200,7 @@ void reconcilable_result_builder::consume_new_partition(const dht::decorated_key
     _static_row_is_alive = false;
     _live_rows = 0;
```
Comment (avikivity) [inline mutation/mutation_partition.cc:2201]: > (cherry picked from commit https://github.com/scylladb/scylladb/commit/36daf43562bb96e044ba3d70edd2c7774c260064)

You should drop the cherry-pick markers, since they don't refer to a commit in this repository. They'll just dangle for most people.

File: service/storage_proxy.cc (hunk 2 above)
Comment (avikivity) [reply service/storage_proxy.cc:4990]: Ok, I think it actually is explicit: all_reached_end().

File: mutation/mutation_partition.cc (hunk 1 above)
Comment (avikivity) [reply mutation/mutation_partition.cc:2222]: This suggestion is out-of-scope for the series, it's a big change. But I want @tgrabiec @denesb @gleb-cloudius opinions on it.

**Comment by avikivity** [comment]
> > Maybe we should make mutation queries return pages without live rows (they wouldn't be empty). Doesn't that fix the problem?
> 
> That is exactly what this PR does.

Not exactly (well I'm not sure at all). I think the coordinator loops trying to get more complete data.

We might drop the loop here. The upper layers can then choose whether to loop or not. cql might loop (say when aggregating) or return a short page, the client might loop or give up due to timeout.

There's some value in looping close to the data source, but if we fetched a pageful of tombstones, getting a page of live rows isn't going to be fast. It's better to give the upper layers a chance to make a decision to stop.

**Comment by avikivity** [comment]
@gleb-cloudius please review

**Comment by denesb** [comment]
> > > Maybe we should make mutation queries return pages without live rows (they wouldn't be empty). Doesn't that fix the problem?
> > 
> > 
> > That is exactly what this PR does.
> 
> Not exactly (well I'm not sure at all). I think the coordinator loops trying to get more complete data.
> 
> We might drop the loop here. The upper layers can then choose whether to loop or not. cql might loop (say when aggregating) or return a short page, the client might loop or give up due to timeout.
> 
> There's some value in looping close to the data source, but if we fetched a pageful of tombstones, getting a page of live rows isn't going to be fast. It's better to give the upper layers a chance to make a decision to stop.

I agree. TBH I am also not completely clear on what the effects of this PR are. I'm still in the process of wrapping my head around it.

File: mutation/mutation_partition.cc (hunk 1 above)
Comment (tgrabiec) [reply mutation/mutation_partition.cc:2222]: That would introduce a regression to scans done on behalf of aggregate queries, since now they would involve a round-trip to the client. We could reduce the impact by sending them not at internal boundary, but time boundary related to the timeout. 

I don't know how that fits with the new forward_service.

**Comment by tgrabiec** [comment]
> We might drop the loop here. The upper layers can then choose whether to loop or not. cql might loop (say when aggregating) or return a short page, the client might loop or give up due to timeout.

What does "the loop here" refer to?

**Comment by denesb** [comment]
> > We might drop the loop here. The upper layers can then choose whether to loop or not. cql might loop (say when aggregating) or return a short page, the client might loop or give up due to timeout.
> 
> What does "the loop here" refer to?

I think he means the reconcile() -> reconcile() retries.

**Comment by denesb** [comment]
Looking at the code, I found out that when we generate a retry, we don't do any "paging", we just bump the row limits and redo the entire read on all replicas. So if we missed our target live row count by 1, we will re-read the entire partition(s)

@tgrabiec you once suggested we should use the reader abstraction for reconciliation and instead of retries, we would just request the next chunk of data. I cannot find the issue where you added this comment anymore. It is also possible that I remember it wrong and my mind projects reader onto all problems.

**Comment by gleb-cloudius** [comment]
I do not remember by now why do we require at least one live row in the result and cannot find it. But probably this is because drivers did not handle empty results gratefully (they considered it to be end of the data IIRC). But we probably fixed it with a flag in the paging state (did we). But there were other reason to push unlimited number of tombstones into the result.  Here https://issues.apache.org/jira/browse/CASSANDRA-8933 explains how not doing so in some cases may result in returning outdated data. Our reconciliation algorithm diverged a lot from Cassandra's one, but we need to make sure that this case is handled.

**Comment by denesb** [comment]
> cl=one, no reconciliation, hot cache: 460 ms
> cl=two, reconciling, no paging (before the patch), hot cache: 6.2 s
> cl=two, reconciling, coordinator-side paging (after the patch), hot cache: 24.3 s

After way too much time working on it, I managed to come up with a test to measure this, and in my testing after the patch requires less time then before the patch (both are significantly less than the times measured by @tgrabiec):
* BEFORE (CL=ALL, reconciling): 794.4328852ms
* AFTER (CL=ALL, reconciling): 164.2720556ms

Both cases use the same data shape:
* N1: 1 live row, 1 million dead rows
* N2: 1 million dead rows, 1 live row

It is possible that https://github.com/scylladb/scylladb/pull/11053 have already brought up the BEFORE time such that tombstone paging + incremental reconciliation now performs better than tombstone paging + non-paged reconciliation.
In the BEFORE case, I also see large allocation warnings, of up to 2.6MIB, there is no such warnings in the AFTER case.

@tgrabiec how did you measure the times from your cover letter?

**Comment by avikivity** [comment]
Note that a 2.6MB large allocation can take any amount of time in a server that has been up for a long time, or even fail.

**Comment by denesb** [comment]
v2:
* Testing has finished, the cover-letter was updated accordingly. Since this PR was originally opened, we have completed the v2 transition and we implemented empty replica pages (https://github.com/scylladb/scylladb/pull/11053). Looks like these two additional changes make this PR completely address reconciling reads. There is no additional range-tombstone build-up on the replica and tombstone limits were implemented for regular reads.

**Comment by denesb** [comment]
> I do not remember by now why do we require at least one live row in the result and cannot find it. But probably this is because drivers did not handle empty results gratefully (they considered it to be end of the data IIRC). But we probably fixed it with a flag in the paging state (did we). But there were other reason to push unlimited number of tombstones into the result. Here https://issues.apache.org/jira/browse/CASSANDRA-8933 explains how not doing so in some cases may result in returning outdated data. Our reconciliation algorithm diverged a lot from Cassandra's one, but we need to make sure that this case is handled.

I will check the code to see if this is handled.

**Comment by denesb** [comment]
> > I do not remember by now why do we require at least one live row in the result and cannot find it. But probably this is because drivers did not handle empty results gratefully (they considered it to be end of the data IIRC). But we probably fixed it with a flag in the paging state (did we). But there were other reason to push unlimited number of tombstones into the result. Here https://issues.apache.org/jira/browse/CASSANDRA-8933 explains how not doing so in some cases may result in returning outdated data. Our reconciliation algorithm diverged a lot from Cassandra's one, but we need to make sure that this case is handled.
> 
> I will check the code to see if this is handled.

After considering this, we must already handle this. The bug described above doesn't go away just by reading until the first live row. In the scenario described above, the unsent tombstone was after the live row. This PR changes nothing in that regard.

**Comment by gleb-cloudius** [comment]
> > > I do not remember by now why do we require at least one live row in the result and cannot find it. But probably this is because drivers did not handle empty results gratefully (they considered it to be end of the data IIRC). But we probably fixed it with a flag in the paging state (did we). But there were other reason to push unlimited number of tombstones into the result. Here https://issues.apache.org/jira/browse/CASSANDRA-8933 explains how not doing so in some cases may result in returning outdated data. Our reconciliation algorithm diverged a lot from Cassandra's one, but we need to make sure that this case is handled.
> > 
> > 
> > I will check the code to see if this is handled.
> 
> After considering this, we must already handle this. The bug described above doesn't go away just by reading until the first live row. In the scenario described above, the unsent tombstone was after the live row. This PR changes nothing in that regard.

The bug above is not about reading until first live row, but about sending all tombstones after the last live row in the page and until next live row. This may also cause the page to be unlimited. Does this PR not cover this case?

**Comment by denesb** [comment]
> > > > I do not remember by now why do we require at least one live row in the result and cannot find it. But probably this is because drivers did not handle empty results gratefully (they considered it to be end of the data IIRC). But we probably fixed it with a flag in the paging state (did we). But there were other reason to push unlimited number of tombstones into the result. Here https://issues.apache.org/jira/browse/CASSANDRA-8933 explains how not doing so in some cases may result in returning outdated data. Our reconciliation algorithm diverged a lot from Cassandra's one, but we need to make sure that this case is handled.
> > > 
> > > 
> > > I will check the code to see if this is handled.
> > 
> > 
> > After considering this, we must already handle this. The bug described above doesn't go away just by reading until the first live row. In the scenario described above, the unsent tombstone was after the live row. This PR changes nothing in that regard.
> 
> The bug above is not about reading until first live row, but about sending all tombstones after the last live row in the page and until next live row. This may also cause the page to be unlimited. Does this PR not cover this case?

The above bug was when limit=X and a tombstone after X'th live row on one replica, that other replicas don't have. The logic around the limit is unchanged, this limit just introduces enforcement of the memory limit, when no live row was seen yet. All the other cases, when the page already has a live row, are unchanged.

**Comment by denesb** [comment]
CI slept on the force-push, will kick it now.

File: mutation/mutation_partition.cc (hunk 1 above)
Comment (avikivity) [reply mutation/mutation_partition.cc:2222]: Aggregations (including those by forward_service) are always paged internally.

File: mutation/mutation_partition.cc [hunk 4]
```diff
@@ -2220,7 +2220,7 @@ stop_iteration reconcilable_result_builder::consume(clustering_row&& cr, row_tom
     }
     _live_rows += is_alive;
     auto stop = _memory_accounter.update_and_check(cr.memory_usage(_schema));
```
Comment (avikivity) [inline mutation/mutation_partition.cc:2223]: > For 1M dead rows, freezing takes 130ms, unfreezing takes 500ms.

Note, this means 130ns/row to freeze, 500ns to unfreeze. At 2GHz  / 2 IPC this is 520 instructions to freeze, 2000 instructions to unfreeze. That's ridiculously high for a row that only contains a tombstone. @michoecho

**Comment by avikivity** [comment]
This looks okay but we need correctness tests. We'll also need mixed-version tests.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3045/

https://github.com/scylladb/scylladb/issues/15046

File: mutation/mutation_partition.cc (hunk 1 above)
Comment (denesb) [reply mutation/mutation_partition.cc:2222]: I just realized I forgot to update this commitlog too.

The current version also addresses client-side timeouts, with the exception of aggregates, which are paged internally.
The coordinator itself will not "loop", looking for empty pages anymore, it will send the empty page to the client, just like we do when hitting the tombstone limit during a regular page.

**Comment by denesb** [comment]
> This looks okay but we need correctness tests. We'll also need mixed-version tests.

We do have read-repair dtests, although not a lot.
I will do a manual mixed-version test.

**Comment by denesb** [comment]
v2.1:
* Update commitlogs.
* Dropped cherry-pick markers.

**Comment by avikivity** [comment]
Filed https://github.com/scylladb/scylladb/issues/15051 for CI failure

**Comment by denesb** [comment]
I completed the mixed-cluster testing, I observed no errors. Read-repair on a mixed cluster took the same amount of time as on a "before" cluster, with the same symptoms (large allocations and stalls).

**Comment by avikivity** [comment]
We need read-repair tests that exercise the new boundaries.

**Comment by denesb** [comment]
> We need read-repair tests that exercise the new boundaries.

Ok, I will look into writing such test(s). I have to think about how to scale the dataset down. The test scenario I used to test this (1mil rows), took more than 15 minutes to run, if creating the cluster and writing the dataset is included.

**Comment by avikivity** [comment]
A million rows shouldn't take 15 minutes. 15 minutes = 900 seconds -> 1100 rows/sec.

The python driver has a way to drive queries in parallel (a little like max_concurrent_parallel_for_each) but I forget its name. Alternatively, configure down the limits so you don't need a million rows.

**Comment by denesb** [comment]
> A million rows shouldn't take 15 minutes. 15 minutes = 900 seconds -> 1100 rows/sec.

Yes, I did it with a simple loop and worked around the time by restoring the data from backup on each run.

> The python driver has a way to drive queries in parallel (a little like max_concurrent_parallel_for_each) but I forget its name. Alternatively, configure down the limits so you don't need a million rows.

I will go with the latter route.

**Comment by denesb** [comment]
v3:
* Added two more patches with tracepoints.
* Added correctness dtest: https://github.com/scylladb/scylla-dtest/pull/3539

**Comment by avikivity** [comment]
What would it take to port the test to test.py? This would let it run regularly in debug mode.

**Comment by denesb** [comment]
> What would it take to port the test to test.py? This would let it run regularly in debug mode.

I don't know, I have very little experience with test.py clusters. I can try.

File: test/topology_custom/test_read_repair.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+    """
+    cmdline = ["--hinted-handoff-enabled", "0", "--query-tombstone-page-limit", "1000"]
+    node1 = await manager.server_add(cmdline=cmdline)
+    node2 = await manager.server_add(cmdline=cmdline)
+
+    await wait_for_token_ring_and_group0_consistency(manager, time.time() + 30)
+
+    cql = manager.get_cql()
+
+    cql.execute("CREATE KEYSPACE ks WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 2}")
```
Comment (mykaul) [inline test/topology_custom/test_read_repair.py:74]: nit: we are trying to deprecate SimpleStrategy, would be nice to move away from it in new tests already.

File: test/topology_custom/test_read_repair.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        assert nr_pages > 1
+
+        assert len(res_rows) == len(expected_rows)
+        for row, expected_row in zip(res_rows, expected_rows):
+            for k, v in expected_row.items():
+                assert getattr(row, k) == v
+
+    logger.info("Check rows with CL=ONE before read-repair")
+    stmt = SimpleStatement(select_query, consistency_level=ConsistencyLevel.ONE)
+    async with connection(manager, node1) as cql:
```
Comment (kbr-scylla) [inline test/topology_custom/test_read_repair.py:112]: Is having those separate connections necessary?

We had to do a lot of fine tuning of Python driver settings for the connection that you obtain through `manager.get_cql()`, in order to fix flakiness. Without doing the same fine tuning for these new connections this test is very likely going to be flaky (Python driver timeouts and so on).

But I wonder if you need the separate connections at all. Do you want to execute a query on a particular node? Establishing a new connection like this does not guarantee which node we'll execute query on (the list passed to `Cluster` is only for initial discovery, the driver automatically establishes connections to each node when discovering them.)

In order to query a particular node, you can use `manager.get_cql()` but pass `host` parameter to `execute` or `run_async`. The `host` parameter expects a `Host` object which you can obtain using `wait_for_cql_and_get_hosts`. See e.g. test_cluster_features.py how it uses `get_supported_features`.

File: test/pylib/manager_client.py [hunk 7]
```diff
@@ -52,9 +52,14 @@ async def stop(self):
         """Close driver"""
         self.driver_close()
 
-    async def driver_connect(self, server: Optional[ServerInfo] = None) -> None:
+    async def driver_connect(self, server: Optional[ServerInfo] | List[ServerInfo] = None) -> None:
```
Comment (kbr-scylla) [inline test/pylib/manager_client.py:55]: Do you use this modification?

**Comment by kbr-scylla** [comment]
> Also, how do I run flake on my code? I see you guys have hints for flake, so it is probably used.

I don't think I ever used flake, but I strongly recommend checking type correctness using mypy.

It should be simple enough; to install:
```
pip install mypy
```

to run:
```
mypy --check-untyped-defs test/topology_custom/test_read_repair.py
```

**Comment by mykaul** [comment]
> I don't think I ever used flake, but I strongly recommend checking type correctness using mypy.

```
sudo dnf install -y python-flake8
flake8
```

File: test/pylib/manager_client.py (hunk 7 above)
Comment (denesb) [reply test/pylib/manager_client.py:55]: Not anymore. Will drop.

File: test/topology_custom/test_read_repair.py (hunk 6 above)
Comment (denesb) [reply test/topology_custom/test_read_repair.py:112]: In my new version I updated this to:
```
 18 @contextlib.asynccontextmanager
 19 async def exclusive_connection(manager, node):
 20     try:
 21         profile = ExecutionProfile(request_timeout=30, load_balancing_policy = WhiteListRoundRobinPolicy([node.ip_addr]))
 22         c = Cluster([node.ip_addr], execution_profiles={EXEC_PROFILE_DEFAULT: profile})
 23         s = c.connect()
 24         yield s
 25     finally:
 26         s.shutdown()
```

`manager_client`'s API to obtain new connections is very clunky so I ended up avoiding it as much as I could.

Comment (denesb) [reply test/topology_custom/test_read_repair.py:112]: I need an exclusive connection to a single node, to verify that each has only the data it's supposed to have, without having to stop the other node (which is slow).

**Comment by denesb** [comment]
v5:
* Clear new test file with `flake8` and `mypy`.
* Drop unneeded change in `manager_client.py`.
* Add additional random rows to each node, to better exercise repair.
* Hopefully fix the failures seen in CI.

**Comment by denesb** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3496/

https://github.com/scylladb/scylladb/issues/15057

But also my new tests are failing in debug mode. I can reproduce it locally too, they always fail in debug mode.

**Comment by denesb** [comment]
I managed to stabilize the test by sacrificing speed. But after increasing the amount of random rows generated, it is flaky in all modes.

File: test/topology_custom/test_read_repair.py (hunk 6 above)
Comment (kbr-scylla) [reply test/topology_custom/test_read_repair.py]: > manager_client's API to obtain new connections is very clunky so I ended up avoiding it as much as I could.

My point is, you don't need new connections. You can use the single connection you already have (i.e. the one given by `manager.get_cql()`)

> I need an exclusive connection to a single node, to verify that each has only the data it's supposed to have, without having to stop the other node (which is slow).

You don't need an exclusive connection for that -- you can

Comment (kbr-scylla) [reply test/topology_custom/test_read_repair.py]: > In my new version I updated this to:

Look into test/topology/conftest.py, `cluster_con` function, all the timeout related parameters there and comments.
You should do the same adjustments.

But:
- you can use `cluster_con` and then the `whitelist` profile when doing queries so it sends the query to the node you initially specified as a contact point
- OR as I would prefer -- use the connection that `manager` already has, passing `host` parameter when executing queries.

**Comment by kbr-scylla** [comment]
>  But after increasing the amount of random rows generated, it is flaky in all modes.

What's the reason of flakiness?

File: test/topology_custom/test_read_repair.py (hunk 6 above)
Comment (denesb) [reply test/topology_custom/test_read_repair.py]: Well actually neither works.
* Using a new exclusive connection doesn't work for the partition-tombstone test-case: range scans on distribued tables can still decide to contact other nodes, there is nothing the client can  do to prevent that.
* Stopping one node and reading from the other also doesn't work, for some reason, some of the time rows are missing from the result.
* Passing `host=node_ip` to `session.execute()` just doesn't work at all. It always fails with `NoHostAvailable`, even if I reconnect the driver one line before trying it.

I have one more approach to try before I give up and just go back to dtests. This is just incredibly painful and frustrting.

Comment (kbr-scylla) [reply test/topology_custom/test_read_repair.py]: > As I explained in https://github.com/scylladb/scylladb/pull/14923#discussion_r1315637682, you need to pass an object of the driver's Host type to the host param. You can translate ServerInfos (node1 or node2) to Hosts using wait_for_cql_and_get_hosts.

Example from test_fencing.py:
```python3
    host2 = (await wait_for_cql_and_get_hosts(manager.cql, [servers[2]], time.time() + 60))[0]

    logger.info(f"trying to write through host2 to regular column [{host2}]")
    with pytest.raises(WriteFailure, match="stale topology exception"):
        await manager.cql.run_async("insert into t1(pk, int_c) values (1, 1)", host=host2)
```

Comment (denesb) [reply test/topology_custom/test_read_repair.py]: The driver silently accepted an ip address too and then said the node is down.

But anyway I am not trying a new approach, which doesn't rely on stopping/starting nodes and trying to control which node a query executes on. The approaches above (exclusive query, passing something to host) only work for single-page/non-paged queries, as soon as you have hundreds of pages, scylla can and does decide to send requests to other replicas as well. Especially if range-scans are involved.

Comment (denesb) [reply test/topology_custom/test_read_repair.py]: > But how would moving to dtests solve these problems? They're using the exact same Python driver.

Yes, writing tests which test repair are just frustrating, regardless of the framework. I expected it to be less painful if I move away from dtest and it is not. I said I will move back to dtest because the test was reliable working there already, it has become flaky after the porting. But now I remembered that I also added change after porting and those would have possibly made the test flaky with dtests too.

Comment (gleb-cloudius) [reply test/topology_custom/test_read_repair.py]: > > The driver silently accepted an ip address too and then said the node is down.
> 
> Yup. That's the Python driver for you. :(

But didn't we decide that typeless languages are great the other day?

Comment (denesb) [reply test/topology_custom/test_read_repair.py]: > > The driver silently accepted an ip address too and then said the node is down.
> 
> Yup. That's the Python driver for you. :(

Well, I saw it in the docs that I should pass a `cassandra.Pool.Host` or something, but like a good Eastern European I thought "surely this speed-limit only applies to cyclists" and passed an ip anyway.

Comment (denesb) [reply test/topology_custom/test_read_repair.py]: > > > The driver silently accepted an ip address too and then said the node is down.
> > 
> > 
> > Yup. That's the Python driver for you. :(
> 
> But didn't we decide that typeless languages are great the other day?

Don't spread the flames. :)) :fire:

**Comment by denesb** [comment]
> > But after increasing the amount of random rows generated, it is flaky in all modes.
> 
> What's the reason of flakiness?

Range scans read from all replicas, ruining my checks for which node has what. I solved this, see below.

**Comment by denesb** [comment]
v6:
* Completely revamp the test: 
     - use `scylla sstable write` + `nodetool refresh` to upload data to nodes, instead of relying on slow stopping+starting the nodes.
     - use `select * from mutation_fragments()` to check which node has what data. This is 100% reliable, without exclusive connections and trying to stop/start nodes.
* Fixed a bug in `select * from mutation_fragments()` paging logic (discovered by the new test).

**Comment by denesb** [comment]
@avikivity the test now also generates 32 random live rows for each node, so we can stress the logic a bit more. I opted for this, instead of trying to nail page boundaries manually. The reason is: if I either fail to generate rows on the page boundary or a later change makes the logic faulty, the test will silently pass. With random rows, there is always a chance that an interesting case will be hit.

**Comment by avikivity** [comment]
>   * use `select * from mutation_fragments()` to check which node has what data. This is 100% reliable, without exclusive connections and trying to stop/start nodes.


Ah, the secret plan is finally laid bare.

**Comment by nyh** [comment]
>       * use `select * from mutation_fragments()` to check which node has what data. This is 100% reliable, without exclusive connections and trying to stop/start nodes.

Nice. I'm writing now a tool to scan a table for materialized-view inconsistencies, and I need to be able to read all the data that one particular replica has - without risking that the scan will switch to another node in the middle. I'll be using the Rust driver (if it makes a different). Is this impossible to do without mutation_fragements? (I guess I'll know better when I implement it and test what it really does, but just wanted to comment that the ability to read only from a single replica is important, especially for testing, and we should have had a reliable and official way to do it)

**Comment by denesb** [comment]
> > ```
> >   * use `select * from mutation_fragments()` to check which node has what data. This is 100% reliable, without exclusive connections and trying to stop/start nodes.
> > ```
> 
> Nice. I'm writing now a tool to scan a table for materialized-view inconsistencies, and I need to be able to read all the data that one particular replica has - without risking that the scan will switch to another node in the middle. I'll be using the Rust driver (if it makes a different). Is this impossible to do without mutation_fragements? (I guess I'll know better when I implement it and test what it really does, but just wanted to comment that the ability to read only from a single replica is important, especially for testing, and we should have had a reliable and official way to do it)

To the best of my knowledge, currently there is no way to restrict scylla to execute a range scan only on a certain replica. It kind-of works for single-partition queries, but it is not reliable (due to e.g. speculative retry).
There is a proposal to open a separate CQL port, which serves all requests from just the local replica, see https://github.com/scylladb/scylladb/pull/15346.

**Comment by denesb** [comment]
I reverted this because the new test is flaky, often breaking CI.

**Comment by denesb** [comment]
See https://github.com/scylladb/scylladb/issues/15371.
Will open a replacement PR soon.

**Comment by nyh** [comment]
> See #15371. Will open a replacement PR soon.

I lost track, did you get a new PR in? If not, we should reopen #9111, right?

**Comment by denesb** [comment]
> > See #15371. Will open a replacement PR soon.
> 
> I lost track, did you get a new PR in? If not, we should reopen #9111, right?

The replacement is https://github.com/scylladb/scylladb/pull/15414, it was queued, then de-queued because of problems with the tests. I will re-open #9111.
